File: packages/cache-go/src/utils/crc32.go
package utils

var crcTable [256]uint32

func init() {
	for i := range 256 {
		c := uint32(i)
		for range 8 {
			if c&1 != 0 {
				c = 0xedb88320 ^ (c >> 1)
			} else {
				c = c >> 1
			}
		}
		crcTable[i] = c
	}
}

func CRC32(data []byte) uint32 {
	crcValue := uint32(0xffffffff)
	for _, b := range data {
		crcValue = crcTable[(crcValue^uint32(b))&0xff] ^ (crcValue >> 8)
	}
	return crcValue ^ 0xffffffff
}

File: packages/cache-go/src/utils/crc32.ts
const crcTable = new Uint32Array(256)

for (let i = 0; i < 256; i++) {
  let c = i
  for (let j = 0; j < 8; j++) {
    c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1
  }
  crcTable[i] = c
}

function crc32(str: string | Uint8Array): string {
  const bytes = typeof str === 'string' ? new TextEncoder().encode(str) : str
  let crcValue = 0xffffffff
  for (let i = 0; i < bytes.length; i++) {
    const byte = bytes[i] ?? 0
    crcValue = (crcTable[(crcValue ^ byte) & 0xff] ?? 0) ^ (crcValue >>> 8)
  }
  return ((crcValue ^ 0xffffffff) >>> 0).toString(16).padStart(8, '0')
}

export default crc32
